rkt: Add GetPodStatus(), GetPodStatusAndAPIPodStatus().

yifan-gu opened the pull request:
Also add ConvertPodStatusToAPIPodStatus().
Add tests for GetPodStatus().

cc @yujuhong @Random-Liu @dgonyeo

Commit:
rkt: Add GetPodStatus(), GetPodStatusAndAPIPodStatus().

Also add ConvertPodStatusToAPIPodStatus().
Add tests for GetPodStatus().

Random-Liu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+
+	// Cannot find any rkt pod associated with the api pod UID.
+	if latestPod == nil {
+		return nil, nil
```
@yifan-gu @yujuhong Should we return an empty pod status here or nil? In docker, we always return a pod status even without any container statuses inside. Which is better?

yujuhong replied:
I think we should be consistent and return an empty pod status, so that the caller can handle them properly.

yifan-gu replied:
ack

Random-Liu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+			StartedAt: latestCreationTime,
+			ExitCode:  int(app.ExitCode),
+			Image:     app.Image.Name,
+			ImageID:   app.Image.Id,
```
@yifan-gu @yujuhong In docker, the image id is added with a prefix "docker://" by kubelet. What is the image id like in rkt? Should we also add prefix to show image type? :)

yifan-gu replied:
ack

Random-Liu replied:
Just curious, does rkt also use docker image now in kubelet?

yifan-gu replied:
@Random-Liu yes, it's using docker image.

Random-Liu replied:
@yujuhong Then, should we use "docker://" instead of "rkt://" here?

yifan-gu replied:
@Random-Liu I think the image id is not related to image formats here.

Random-Liu replied:
In fact, I don't quite understand the prefix here, is it just a namespace or something, :) So just make sure.

yifan-gu replied:
@Random-Liu I think it's just a namespacing thing.

yifan-gu replied:
@Random-Liu Maybe we should add the prefix in the kubelet level when transforming the raw pod status to api pod status. I will add the prefix now though.

Random-Liu replied:
Yeah, I think you are right, :) I checked the code, no one relies on it. Even in the api [document](http://kubernetes.io/v1.1/docs/api-reference/v1/definitions.html), there is nothing about that prefix. I believe it is just a name, haha

Random-Liu replied:
@yifan-gu Ok, thanks! :0

yujuhong commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+		Namespace: namespace,
+	}
+
+	// TODO(yifan): Use ListPods(detail=true) to avoid another InspectPod rpc here.
```
The container runtime interface we have now is more docker-centric, so PLEG does GetPods() to detect changes and then inspect the pod for details. We should monitor the resource usage of rkt to see if this flow/interface work well with rkt for our scalability goals.

yujuhong commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+		}
+
+		creationTimeStr, ok := manifest.Annotations.Get(k8sRktCreationTimeAnno)
+		if !ok {
```
In what cases would this be true and we won't have a creation timestamp? If this is an old pod that doesn't have the timestamp, I feel like we should still include the pod in the results.

yujuhong commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+			}
+		}
+	}
+
```
The main concern I have here is:
 - In what cases would the annotations be missing/misformatted?
 - Should we fail in such cases?

For docker, this usually implies that the container was started before the kubelet upgrade, and we should try accommodating them (and not fail). Shouldn't we do the same for rkt?

yifan-gu replied:
@yujuhong One case is what you mentioned, the kubelet is update. But other cases are possible as well, because the annotations are just strings in a file on the disk, which is always possible to be modified. (currently, only root users are able to change them).

yujuhong replied:
What I am worried about is that, without properly including those pods, kubelet may falsely restart a new pod. Of course, in the case of user errors, we might be able to get away with it (i.e., user's fault). For other cases, maybe it's worth returning the pod with empty field? What do you think.

yifan-gu replied:
@yujuhong I think in this PR, as all the pods are fetched via rkt api service, and we already assume them to contain those annotations (that's how we filter the pods), so it's already a breaking change. So under that assumption, the annotations I am trying to find here, should always exist unless there are user errors.

yujuhong replied:
Ack

yujuhong commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+				latestPod = inspectResp.Pod
+				latestPodManifest = &manifest
+				latestCreationTime = time.Unix(unixSec, 0)
+			}
```
Hm....GetPodStatus is supposed to return all container statuses (incl. the dead containers). You actually inspected all of dead pods, but instead of returning all of them, you only return the latest one.

yifan-gu replied:
@yujuhong We slightly discussed this In https://github.com/kubernetes/kubernetes/pull/17420#discussion_r46497969 . This really is the part where the "pod vs container" thing gets into problems. I think I can work around by returning a pod which combines all apps from previous pods, do you think it will be better?

yujuhong replied:
It might be necessary evil for now so that the output is consistent across runtimes.

yifan-gu replied:
@yujuhong But the output is already different for things like GetPods(). The pod returned by GetPods() for rkt will contain exactly the same number of containers as in the pod spec, however it's not the case in docker.

yujuhong replied:
Hmm....if rkt doesn't return this, it'd be hard for kubelet to compute the last termination state, etc. CAn we make it consistent?

yifan-gu replied:
@yujuhong I suggest we keep consistent within a runtime? 
As the pod level interfaces are created for pod-level runtimes like rkt. The implementation path we took for docker is just because it lacks "pod", there's no way to figure out whether a dead container belongs to an "older" pod or an "older than older" pod. That's why we have more containers in a pod returned by GetPods() than the pod spec for docker. 

On the other hand, rkt already has the pod concept, it's looks like a backfire if we choose the docker path for this kind of interface when integrating rkt.

yujuhong replied:
Yes and no. Unfortunately, we also rely on seeing dead containers to populate the api.PodStatus object. Suppose rkt restarts the pod in SyncPod(), how would kubelet know the LastTerminationState? If we can split SyncPod into concrete steps, we may be able to work around this, but I don't see it happening anytime soon...

yujuhong replied:
Another potential benefit of returning everything is that kubelet might be able to determine what's eligible for garbage collection in the future.

yifan-gu replied:
@yujuhong I don't understand why kubelet cannot know the LastTerminationState?

yujuhong replied:
I don't know if overloading SyncPod's returned object would be a good idea... Let's walk through an example:
- In SyncPod()
 * kill container A (id: a1)
 * start container A (id: a2)

Right now, the pleg would detect the container changes and populate the runtime cache with the PodStatus, which includes both container a1 and a2.

If we don't do that, SyncPod needs to inspect a1 between the two operations and return the container status as part of the result. We then have to handle this returned object to populate the last termination state. Overall it's more complicated to deal with.

Random-Liu commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
+			ContainerID:  c.ID.String(),
+			RestartCount: c.RestartCount,
+			State:        st,
+		}
```
@yifan-gu @yujuhong LastTerminationState hasn't been set here. What is LastTermination used for? Just for introspection or there is some functionality relies on it?

yifan-gu replied:
OOB discussed with @yujuhong , thought the LastTermination should be implemented at kubelet level.

Random-Liu replied:
Ok. :)

Commit:
rkt: Fix GetPods(), refactor tests for GetPods().

Fix GetPods() so that the container hash is fetched from the
annotations in pod manifest's app list instead of image manifest.

yifan-gu commented:
@dgonyeo I added another commit to fix the container hash retrieval part in GetPods(), also refactoring the tests using the helper functions to generate rktPod.

yifan-gu commented:
@yujuhong @Random-Liu Changed GetPodStatus() to return dead container status as well.

yifan-gu commented:
cc @jonboulle

yujuhong commented:
LGTM. Thanks!

Random-Liu commented:
LGTM. Thanks a lot! @yifan-gu
